datamade/how-to
# Add browserify-css and some CSS in Django documentation; tiny cookiecutter fixes

## Overview
This PR:
- Closes #236 
  - Adds the browserify-css library to work with our asset pipeline
  - Includes documentation about organizing CSS

Includes:
- Fix to the docker-compose.yml that prevented the cookiecutter from creating a django app
- Fix to the run_tests.sh script so the cookiecutter creates a test command with the app name

## Background
I found that third-party CSS isn't compiled whenever the CSS is imported from node_modules, because our pipeline is configured to only import javascript files from node_modules. One example for this bug is react-leaflet or rc-slider — using these libraries in React code requires you to import a CSS file into the React code. This importing breaks with our current pipepline. See the steps to reproduce in #236 if you want to reproduce this.

### Notes
I added the `browserify-css` library as a **devDependency** in `package.json`. This is the only way I could get this to work within the cookiecutter. I tried adding it as a regular dependency to the cookiecutter, but the library wouldn't download whenever the docker image built after creating the app via cookiecutter. I have no idea why — this was quite frustrating to debug. Perhaps this is because [the library documentation](https://www.npmjs.com/package/browserify-css) says to `npm install --save-dev browserify-css`?

Packaging this as a "devDependency" seems like it wouldn't work when deployed to heroku, but somehow it still does. You can see my demo app as well as this screenshot where the third party CSS was packaged.

- working slider that uses css from node_modules: https://test-css-staging.herokuapp.com/
- screenshot that shows the slider and the css:
<img width="866" alt="Screen Shot 2022-06-16 at 1 10 44 PM" src="https://user-images.githubusercontent.com/38969506/174161085-be833d51-b73e-45b8-a815-19a0e60d0c18.png">

You can also shell into the heroku app and see the library in node_modules. You can see other libraries from "devDependencies" in node_modules, like eslint.



## Testing Instructions
- Create new app with cookiecutter.
- Create a basic React component and render it on a page.
- Download an npm package that requires you to import their css. [rc-slider is a quick and easy one](https://www.npmjs.com/package/rc-slider).
- Import the CSS into the React component.
- Make sure the CSS works and the app doesn't crash when you visit the page with the CSS.
- Make sure you can see the CSS in the HTML header.
- For good measure, you can deploy the test app to heroku and make sure it works as expected.
- Review documentation for clarity and correctness.
